chips/lowrisc/src/i2c.rs

```diff
@@ -0,0 +1,494 @@
+    FIFO_CTRL [
+        RXRST OFFSET(0) NUMBITS(1) [],
+        FMTRST OFFSET(1) NUMBITS(1) [],
+        RXILVL OFFSET(2) NUMBITS(3) [
+            RXLVL1 = 0,
+            RXLVL4 = 1,
+            RXLVL8 = 2,
+            RXLVL16 = 3,
+            RXLVL30 = 4
+        ],
+        FMTILVL OFFSET(5) NUMBITS(3) [
+            FMTLVL1 = 0,
+            FMTLVL4 = 1,
+            FMTLVL8 = 2,
+            FMTLVL16 = 3,
+            FMTLVL30 = 4
+        ]
+    ],
+    FIFO_STATUS [
+        FMTLVL OFFSET(0) NUMBITS(6) [],
+        RXLVL OFFSET(16) NUMBITS(6) []
+    ],
+    OVRD [
+        TXOVRDEN OFFSET(0) NUMBITS(1) [],
+        SCLVAL OFFSET(1) NUMBITS(1) [],
+        SDAVAL OFFSET(2) NUMBITS(1) []
+    ],
+    VAL [
+        SCL_RX OFFSET(0) NUMBITS(1) [],
+        SDA_RX OFFSET(1) NUMBITS(1) []
+    ],
+    TIMING0 [
+        THIGH OFFSET(0) NUMBITS(16) [],
+        TLOW OFFSET(16) NUMBITS(16) []
+    ],
+    TIMING1 [
+        T_R OFFSET(0) NUMBITS(16) [],
+        T_F OFFSET(16) NUMBITS(16) []
+    ],
+    TIMING2 [
+        TSU_STA OFFSET(0) NUMBITS(16) [],
+        THD_STA OFFSET(16) NUMBITS(16) []
+    ],
+    TIMING3 [
+        TSU_DAT OFFSET(0) NUMBITS(16) [],
+        THD_DAT OFFSET(16) NUMBITS(16) []
+    ],
+    TIMING4 [
+        TSU_STO OFFSET(0) NUMBITS(16) [],
+        T_BUF OFFSET(16) NUMBITS(16) []
+    ],
+    TIMEOUT_CTRL [
+        VAL OFFSET(0) NUMBITS(31) [],
+        EN OFFSET(31) NUMBITS(1) []
+    ]
+];
+
+pub struct I2c<'a> {
+    registers: StaticRef<I2cRegisters>,
+    clock_period_nanos: u32,
+
+    master_client: OptionalCell<&'a dyn hil::i2c::I2CHwMasterClient>,
+
+    // Set whe calling the write_read operation
+    address: Cell<u8>,
+
+    buffer: TakeCell<'static, [u8]>,
+    write_len: Cell<usize>,
+    write_index: Cell<usize>,
+
+    read_len: Cell<usize>,
+    read_index: Cell<usize>,
+}
+
+impl I2c<'a> {
+    pub const fn new(base: StaticRef<I2cRegisters>, clock_period_nanos: u32) -> I2c<'a> {
+        I2c {
+            registers: base,
+            clock_period_nanos,
+            master_client: OptionalCell::empty(),
+            address: Cell::new(0),
+            buffer: TakeCell::empty(),
+            write_len: Cell::new(0),
+            write_index: Cell::new(0),
+            read_len: Cell::new(0),
+            read_index: Cell::new(0),
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+        let irqs = regs.intr_state.get();
+
+        // Clear all interrupts
+        regs.intr_state.modify(
+            INTR::FMT_WATERMARK::SET
+                + INTR::RX_WATERMARK::SET
+                + INTR::FMT_OVERFLOW::SET
+                + INTR::RX_OVERFLOW::SET
+                + INTR::NAK::SET
+                + INTR::SCL_INTERFERENCE::SET
+                + INTR::SDA_INTERFERENCE::SET
+                + INTR::STRETCH_TIMEOUT::SET
+                + INTR::SDA_UNSTABLE::SET,
+        );
+
+        if irqs & (1 << 0) != 0 {
+            // FMT Watermark
```

Comment:
Is there a reason for using bare constants here rather than the register interface?

Reply:
There isn't, I just didn't know how to use the register interface on a stored value.

Reply:
`let irqs = regs.intr_state.extract();`